File: lib/secrets.ts
// lib/secrets.ts — Cloud KMS / Secret Manager integration
//
// Replaces env-based secrets (INTERNAL_API_TOKEN, OTP_PEPPER) with
// cloud-managed secrets from AWS Secrets Manager or GCP Secret Manager.
// Falls back to env vars for local development.
//
// Usage:
//   const token = await getSecret("INTERNAL_API_TOKEN");
//   const pepper = await getSecret("OTP_PEPPER");

import { withClientSpan, addSpanAttributes } from "@/lib/tracing";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

type SecretProvider = "env" | "aws_sm" | "gcp_sm";

interface CachedSecret {
  value: string;
  fetchedAt: number;
  versionId?: string;
}

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

const PROVIDER: SecretProvider =
  (process.env.SECRET_PROVIDER as SecretProvider) || "env";

const CACHE_TTL_MS = 5 * 60 * 1000; // 5-minute cache TTL
const AWS_REGION = process.env.AWS_REGION || "ap-south-1";
const GCP_PROJECT = process.env.GCP_PROJECT_ID || "";

// Secret name mapping: env var name → cloud secret name
const SECRET_MAPPING: Record<string, string> = {
  INTERNAL_API_TOKEN: process.env.SM_INTERNAL_API_TOKEN_NAME || "brainmate/auth/internal-api-token",
  OTP_PEPPER: process.env.SM_OTP_PEPPER_NAME || "brainmate/auth/otp-pepper",
  CLERK_SECRET_KEY: process.env.SM_CLERK_SECRET_NAME || "brainmate/auth/clerk-secret-key",
  REDIS_URL: process.env.SM_REDIS_URL_NAME || "brainmate/auth/redis-url",
};

// ---------------------------------------------------------------------------
// In-memory cache
// ---------------------------------------------------------------------------

const secretCache = new Map<string, CachedSecret>();

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Retrieve a secret by name. Uses cache to avoid excessive cloud API calls.
 * Falls back to env vars if cloud provider is not configured or fails.
 */
export async function getSecret(name: string): Promise<string> {
  return getSecretWithOptions(name, { forceRefresh: false });
}

export async function getSecretWithOptions(
  name: string,
  options: { forceRefresh?: boolean } = {},
): Promise<string> {
  // Check cache first
  const cached = secretCache.get(name);
  if (
    !options.forceRefresh &&
    cached &&
    Date.now() - cached.fetchedAt < CACHE_TTL_MS
  ) {
    return cached.value;
  }

  return withClientSpan("secrets.get", async () => {
    addSpanAttributes({
      "secret.name": name,
      "secret.provider": PROVIDER,
    });

    let value: string | null = null;

    switch (PROVIDER) {
      case "aws_sm":
        value = await fetchFromAWSSecretsManager(name);
        break;
      case "gcp_sm":
        value = await fetchFromGCPSecretManager(name);
        break;
      case "env":
      default:
        value = process.env[name] ?? null;
        break;
    }

    // Fallback to env var if cloud fetch fails
    if (!value) {
      value = process.env[name] ?? "";
      if (PROVIDER !== "env") {
        console.warn(
          `[secrets] Cloud fetch failed for ${name}, using env var fallback`,
        );
      }
    }

    // Update cache
    secretCache.set(name, {
      value,
      fetchedAt: Date.now(),
    });

    return value;
  }, {
    "rpc.system": "secrets",
    "rpc.method": "getSecret",
  });
}

/**
 * Invalidate a cached secret (e.g., after rotation detection).
 */
export function invalidateSecret(name: string): void {
  secretCache.delete(name);
}

/**
 * Invalidate all cached secrets.
 */
export function invalidateAllSecrets(): void {
  secretCache.clear();
}

/**
 * Check if secrets need rotation (version changed in cloud provider).
 * Returns list of secret names that have newer versions available.
 */
export async function checkRotation(): Promise<string[]> {
  if (PROVIDER === "env") return [];

  const rotated: string[] = [];

  for (const [envName] of Object.entries(SECRET_MAPPING)) {
    const cached = secretCache.get(envName);
    if (!cached) continue;

    try {
      const current = await getSecretWithOptions(envName, { forceRefresh: true });
      if (current !== cached.value) {
        rotated.push(envName);
        console.log(`[secrets] Rotation detected for ${envName}`);
      }
    } catch {
      // Skip rotation check on error
    }
  }

  return rotated;
}

// ---------------------------------------------------------------------------
// AWS Secrets Manager
// ---------------------------------------------------------------------------

async function fetchFromAWSSecretsManager(name: string): Promise<string | null> {
  const secretName = SECRET_MAPPING[name] || name;

  try {
    // Use runtime-only dynamic import to keep cloud SDKs optional.
    const awsSdk = await optionalImport("@aws-sdk/client-secrets-manager");
    if (!awsSdk) return null;
    const { SecretsManagerClient, GetSecretValueCommand } = awsSdk as {
      SecretsManagerClient: new (args: { region: string }) => {
        send: (command: unknown) => Promise<{
          SecretString?: string;
          SecretBinary?: Uint8Array;
        }>;
      };
      GetSecretValueCommand: new (args: { SecretId: string }) => unknown;
    };

    const client = new SecretsManagerClient({ region: AWS_REGION });
    const command = new GetSecretValueCommand({ SecretId: secretName });
    const response = await client.send(command);

    if (response.SecretString) {
      // Try to parse as JSON (multi-value secret)
      try {
        const parsed = JSON.parse(response.SecretString);
        // If it's a JSON object, return the value matching the env var name
        if (typeof parsed === "object" && name in parsed) {
          return parsed[name];
        }
      } catch {
        // Not JSON — return raw string
      }
      return response.SecretString;
    }

    if (response.SecretBinary) {
      return Buffer.from(response.SecretBinary).toString("utf-8");
    }

    return null;
  } catch (err) {
    console.error(`[secrets] AWS Secrets Manager error for ${secretName}:`, err);
    return null;
  }
}

// ---------------------------------------------------------------------------
// GCP Secret Manager
// ---------------------------------------------------------------------------

async function fetchFromGCPSecretManager(name: string): Promise<string | null> {
  const secretName = SECRET_MAPPING[name] || name;

  try {
    const gcpSdk = await optionalImport("@google-cloud/secret-manager");
    if (!gcpSdk) return null;
    const { SecretManagerServiceClient } = gcpSdk as {
      SecretManagerServiceClient: new () => {
        accessSecretVersion: (args: { name: string }) => Promise<Array<{
          payload?: { data?: Uint8Array | string };
        }>>;
      };
    };

    const client = new SecretManagerServiceClient();
    const fullName = `projects/${GCP_PROJECT}/secrets/${secretName}/versions/latest`;

    const [response] = await client.accessSecretVersion({ name: fullName });
    const payload = response.payload?.data;

    if (payload) {
      return typeof payload === "string"
        ? payload
        : new TextDecoder().decode(payload as Uint8Array);
    }

    return null;
  } catch (err) {
    console.error(`[secrets] GCP Secret Manager error for ${secretName}:`, err);
    return null;
  }
}

async function optionalImport(moduleName: string): Promise<unknown | null> {
  try {
    const dynamicImport = new Function(
      "m",
      "return import(m)",
    ) as (m: string) => Promise<unknown>;
    return await dynamicImport(moduleName);
  } catch {
    return null;
  }
}

// ---------------------------------------------------------------------------
// Background rotation checker (starts in instrumentation.ts)
// ---------------------------------------------------------------------------

let rotationInterval: ReturnType<typeof setInterval> | null = null;

/**
 * Start periodic rotation checking (every 30 seconds).
 */
export function startRotationChecker(): void {
  if (PROVIDER === "env" || rotationInterval) return;

  rotationInterval = setInterval(async () => {
    try {
      const rotated = await checkRotation();
      if (rotated.length > 0) {
        console.log(`[secrets] Rotated secrets detected: ${rotated.join(", ")}`);
        // Invalidate rotated secrets so next access fetches fresh values
        for (const name of rotated) {
          invalidateSecret(name);
        }
      }
    } catch (err) {
      console.error("[secrets] Rotation check error:", err);
    }
  }, 30_000);

  console.log("[secrets] Rotation checker started");
}

/**
 * Stop the rotation checker.
 */
export function stopRotationChecker(): void {
  if (rotationInterval) {
    clearInterval(rotationInterval);
    rotationInterval = null;
  }
}

File: rust-auth-gateway/src/secrets.rs
// src/secrets.rs — Cloud KMS / Secret Manager integration for Rust gateway
//
// Replaces env-based secrets with cloud-managed secrets.
// Supports AWS Secrets Manager, with env var fallback for local dev.

use std::collections::HashMap;
use std::env;
use std::sync::Arc;
use std::time::{Duration, Instant};

use tokio::sync::RwLock;
use tracing::Instrument;
use tracing::{info, warn};

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq)]
pub enum SecretProvider {
    Env,
    AwsSecretsManager,
    GcpSecretManager,
}

impl SecretProvider {
    pub fn from_env() -> Self {
        match env::var("SECRET_PROVIDER")
            .unwrap_or_else(|_| "env".to_string())
            .to_lowercase()
            .as_str()
        {
            "aws_sm" | "aws" => Self::AwsSecretsManager,
            "gcp_sm" | "gcp" => Self::GcpSecretManager,
            _ => Self::Env,
        }
    }
}

struct CachedSecret {
    value: String,
    fetched_at: Instant,
}

// ---------------------------------------------------------------------------
// Secret Manager
// ---------------------------------------------------------------------------

pub struct SecretManager {
    provider: SecretProvider,
    cache: Arc<RwLock<HashMap<String, CachedSecret>>>,
    cache_ttl: Duration,
    secret_mapping: HashMap<String, String>,
}

impl Default for SecretManager {
    fn default() -> Self {
        Self::new()
    }
}

impl SecretManager {
    pub fn new() -> Self {
        let mut mapping = HashMap::new();
        // Map env var names to cloud secret names
        mapping.insert(
            "INTERNAL_API_TOKEN".to_string(),
            env::var("SM_INTERNAL_API_TOKEN_NAME")
                .unwrap_or_else(|_| "brainmate/auth/internal-api-token".to_string()),
        );
        mapping.insert(
            "OTP_PEPPER".to_string(),
            env::var("SM_OTP_PEPPER_NAME")
                .unwrap_or_else(|_| "brainmate/auth/otp-pepper".to_string()),
        );
        mapping.insert(
            "REDIS_URL".to_string(),
            env::var("SM_REDIS_URL_NAME")
                .unwrap_or_else(|_| "brainmate/auth/redis-url".to_string()),
        );

        Self {
            provider: SecretProvider::from_env(),
            cache: Arc::new(RwLock::new(HashMap::new())),
            cache_ttl: Duration::from_secs(300), // 5 minutes
            secret_mapping: mapping,
        }
    }

    /// Get a secret by env var name. Uses cache, falls back to env var.
    pub async fn get_secret(&self, name: &str) -> String {
        // Check cache
        {
            let cache = self.cache.read().await;
            if let Some(cached) = cache.get(name) {
                if cached.fetched_at.elapsed() < self.cache_ttl {
                    return cached.value.clone();
                }
            }
        }

        // Fetch from provider
        let value = match self.provider {
            SecretProvider::AwsSecretsManager => self.fetch_aws(name).await.unwrap_or_else(|| {
                warn!("AWS SM fetch failed for {name}, falling back to env");
                env::var(name).unwrap_or_default()
            }),
            SecretProvider::GcpSecretManager => self.fetch_gcp(name).await.unwrap_or_else(|| {
                warn!("GCP SM fetch failed for {name}, falling back to env");
                env::var(name).unwrap_or_default()
            }),
            SecretProvider::Env => env::var(name).unwrap_or_default(),
        };

        // Update cache
        {
            let mut cache = self.cache.write().await;
            cache.insert(
                name.to_string(),
                CachedSecret {
                    value: value.clone(),
                    fetched_at: Instant::now(),
                },
            );
        }

        value
    }

    /// Invalidate a specific cached secret.
    pub async fn invalidate(&self, name: &str) {
        let mut cache = self.cache.write().await;
        cache.remove(name);
    }

    /// Start background rotation checker.
    pub fn spawn_rotation_checker(self: Arc<Self>) {
        if self.provider == SecretProvider::Env {
            return;
        }

        tokio::spawn(
            async move {
                let mut interval = tokio::time::interval(Duration::from_secs(30));
                interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);

                loop {
                    interval.tick().await;

                    let names: Vec<String> = {
                        let cache = self.cache.read().await;
                        cache.keys().cloned().collect()
                    };

                    for name in names {
                        let old_value = {
                            let cache = self.cache.read().await;
                            cache.get(&name).map(|c| c.value.clone())
                        };

                        if let Some(old) = old_value {
                            self.invalidate(&name).await;
                            let new_value = self.get_secret(&name).await;
                            if new_value != old {
                                info!(secret = %name, "secret rotation detected");
                                metrics::counter!("secret_rotation_detected_total").increment(1);
                            }
                        }
                    }
                }
            }
            .instrument(tracing::info_span!("secret_rotation_checker")),
        );
    }

    // AWS Secrets Manager fetch
    async fn fetch_aws(&self, name: &str) -> Option<String> {
        let secret_name = self
            .secret_mapping
            .get(name)
            .cloned()
            .unwrap_or_else(|| name.to_string());

        #[cfg(feature = "aws-secrets")]
        {
            use aws_sdk_secretsmanager::Client;
            let sdk_config = aws_config::load_defaults(aws_config::BehaviorVersion::latest()).await;
            let client = Client::new(&sdk_config);
            return client
                .get_secret_value()
                .secret_id(&secret_name)
                .send()
                .await
                .ok()
                .and_then(|r| r.secret_string().map(|s| s.to_string()));
        }

        #[cfg(not(feature = "aws-secrets"))]
        {
            warn!(
                secret = %secret_name,
                "AWS Secrets Manager requested but aws-secrets feature not enabled. Add --features aws-secrets to Cargo build args."
            );
            None
        }
    }

    // GCP Secret Manager fetch (placeholder)
    async fn fetch_gcp(&self, name: &str) -> Option<String> {
        let secret_name = self
            .secret_mapping
            .get(name)
            .cloned()
            .unwrap_or_else(|| name.to_string());

        #[cfg(feature = "gcp-secrets")]
        {
            use google_secretmanager1::SecretManager;
            // Implementation requires the google-secretmanager1 crate.
            // Add to Cargo.toml: google-secretmanager1 = { version = "5", optional = true }
            // Then compile with --features gcp-secrets
            // Full implementation:
            //   let hub = SecretManager::new(...).await;
            //   let name = format!("projects/{}/secrets/{}/versions/latest", project_id, secret_name);
            //   hub.projects().secrets_versions_access(&name).doit().await...
            let _ = SecretManager::new;
            let _ = secret_name;
            warn!("GCP Secret Manager feature compiled in but implementation incomplete");
            return None;
        }

        #[cfg(not(feature = "gcp-secrets"))]
        {
            warn!(
                secret = %secret_name,
                "GCP Secret Manager requested (SECRET_PROVIDER=gcp_sm) but gcp-secrets \
                 feature is not enabled. To use GCP Secret Manager: \
                 1) Add google-secretmanager1 to Cargo.toml, \
                 2) Add [features] gcp-secrets = [\"google-secretmanager1\"], \
                 3) Implement fetch_gcp(), \
                 4) Compile with --features gcp-secrets. \
                 Falling back to INTERNAL env var for: {}",
                name
            );
            None
        }
    }
}
